### ISSUE TITLE
How to load a new map in Carla and Apollo?

### ISSUE THREAD

**Issue by LCaptainJackSparrow** [guardstrikelab/carla_apollo_bridge#158]
1. Create a road network using RoadRunner
2. I can load new maps in Apollo's dreamview, but I cannot load new maps in Carla
3. How to slove the problem?

**Comment by XiaoFei9704** [comment]
As I know, if you want to load a new opendrive map in Carla, you need everything to construct the world. For example, the assets like buildings, pavement materials, trees, etc...
So it takes much more in Carla to create a new map instead of a single xodr file.

**Comment by LCaptainJackSparrow** [comment]
Yes, I  builded CARLA from the source code for development purposes.I obtained these assets through “make import” in terminal. 
![2023-09-21 16-45-43屏幕截图]
![2023-09-21 16-44-49屏幕截图](https://github.com/guardstrikelab/carla_apollo_bridge/assets/101320438/8b209666-2b7d-4370-99ae-0f0b24aecb51)
(https://github.com/guardstrikelab/carla_apollo_bridge/assets/101320438/7ff08339-c81b-464b-8def-0abfbc4c6b5d)

But I don't know how to use it in this bridge.

**Comment by XiaoFei9704** [comment]
Can you now load your new map in Carla?
If you don't, you may take some more time to tackle it, I can't help you with this cause I'm not familiar with it.
If you do, you'd better convert the Carla map to Apollo using other utils rather than using the map generated by RoadRunner cause there're some coordinate problems with it and they're hard to tackle.
I recommend a map converting util: https://github.com/daohu527/imap  
We've been working with this util for several months, and all maps of this project were generated using this util (repo).

**Comment by LCaptainJackSparrow** [comment]
No, I can't.   Thank you for your reply. I will try the tool you recommended tomorrow

**Comment by XiaoFei9704** [comment]
Ok, that's fine.
But I'd mention that importing a new map to Carla is a big project. The steps are:
1. Create your map by Roadrunner and export it as an opendrive file
2. Prepare any assets the map needs such as buildings and trees
3. Construct the world in UnrealEngine using your map - **Big Step**
4. Convert Carla map to Apollo map
5. Run co-simulation

**Comment by LCaptainJackSparrow** [comment]
I still can't load new maps. I wonder if you can provide more detailed steps. I think there is a missing step between step 3 and step 4, but I am not sure. I have another question. Can the map replaced by co-simulation only be named as **carla_ townx**。 To better analyze the problem, my operation is as follows:

1.Create your map by Roadrunner and export it as an opendrive file，**test.xodr**
2.Prepare any assets the map needs such as buildings and trees
3. Construct the world in UnrealEngine using your map - **Big Step**. After generating the map_package, **I don't know how to load the new map in carla_apollo_bridge**. I got a map_package. It contains many files, as shown in the figure below.
![2023-09-22 11-29-54屏幕截图](https://github.com/guardstrikelab/carla_apollo_bridge/assets/101320438/79ace3b7-8c1c-4117-9ffb-2095237df68a)

4.Convert Carla map to Apollo map  (generate through the link you shared: https://github.com/daohu527/imap)。 I got test.bin and test.txt in this step. Sim_map.bin and sim_map.bin were generated based on these test_map.
![2023-09-22 11-32-36屏幕截图](https://github.com/guardstrikelab/carla_apollo_bridge/assets/101320438/7e63d788-8b11-4c71-8399-c5fcb0a45124)

5.Run co-simulation。Before this step, I changed the 'town:Town01' in the apollo/modules/carla_bridge/config/settings.yaml  to 'town:test' .And I copied the apollo/modules/map/data/test  to /apollo/modules/carla_bridge/map/test

As a result, the newly loaded map can be seen in Dreamview, but not in Carlaue. The relevant error messages are as follows:

![2023-09-22 11-37-49屏幕截图](https://github.com/guardstrikelab/carla_apollo_bridge/assets/101320438/0e2ef36d-2ed6-4975-8f4a-59c87a16e4a6)

**Comment by LCaptainJackSparrow** [comment]
I can load the map now.  I didn't understand the difference between the compiled version and the non-compiled version of Carla before. Thank you very much for your response.

**Comment by congcongla** [comment]
如果你仅仅需要道路，而不需要其他建筑等。通过测试，使用carla提供的generate_opendrive_world能更方便解决这个问题。使用opendrive创建的world和通过make import 创建的场景。在道路的使用上几乎没有区别，只是没有额外的内容，比如道路上的黄白线。
